packages/cdktf/lib/terraform-functions.ts

```diff
@@ -0,0 +1,1250 @@
+import { Tokenization } from "./tokens/token";
+import { call } from "./tfExpression";
+import { IResolvable } from "./tokens/resolvable";
+import { TokenMap } from "./tokens/private/token-map";
+import { TokenString } from "./tokens/private/encoding";
+
+// We use branding here to ensure we internally only handle validated values
+// this allows us to catch usage errors before terraform does in some cases
+type TFValue = any & { __type: "tfvalue" };
+type TFValueValidator = (value: any) => TFValue;
+
+type ExecutableTfFunction = (...args: any[]) => IResolvable;
+
+// Validators
+function anyValue(value: any): any {
+  return value;
+}
+
+function mapValue(value: any): any {
+  return value;
+}
+
+function stringValue(value: any): any {
+  if (typeof value !== "string" && !Tokenization.isResolvable(value)) {
+    throw new Error(`${value} is not a valid number nor a token`);
+  }
+  return value;
+}
+
+function numericValue(value: any): any {
+  if (typeof value !== "number" && !Tokenization.isResolvable(value)) {
+    throw new Error(`${value} is not a valid number nor a token`);
+  }
+  return value;
+}
+
+function listOf(type: TFValueValidator): TFValueValidator {
+  return (value: any) => {
+    if (Tokenization.isResolvable(value)) {
+      return value;
+    }
+
+    if (!Array.isArray(value)) {
+      //   throw new Error(`${value} is not a valid list`);
+      return value;
+    }
+
+    return value.map((item, i) => {
+      if (Tokenization.isResolvable(item)) {
+        return item;
+      }
+
+      if (TokenString.forListToken(item).test()) {
+        return item;
+      }
+
+      if (typeof item === "string") {
+        const tokenList = Tokenization.reverseString(item);
+        const numberOfTokens =
+          tokenList.tokens.length + tokenList.intrinsic.length;
+        if (numberOfTokens === 1 && tokenList.literals.length === 0) {
+          return item;
+        }
+      }
+
+      try {
+        type(item);
+        return typeof item === "string" ? `"${item}"` : item;
+      } catch (error) {
+        throw new Error(
+          `Element in list ${value} at position ${i} is not of the right type: ${error}`
+        );
+      }
+    });
+  };
+}
+
+// Tokenization
+function asString(value: IResolvable) {
+  return TokenMap.instance().registerString(value);
+}
+
+function asNumber(value: IResolvable) {
+  return TokenMap.instance().registerNumber(value);
+}
+
+function asList(value: IResolvable) {
```

Comment:
Since this always returns `string[]` regardless of the type, using it for functions that return other types of lists will give back a not very useable array.

Reply:
I mean this is the current way how we represent lists, isn't it?

Reply:
It is, so could definitely be tackled as a followup.